Streamlining

ccutrer opened the pull request:
Streamline several aspects of the communication between the controller and us.

If you have hundreds of accessories, it's important to not be so chatty.

timcharper committed:
Strip back window covering to essential characteristics

This makes it work once again

timcharper committed:
Bugfix: TargetHeatingMode and CurrentHeatingMode disparity

TargetHeatingMode has 4 enum values, while CurrentHeatingMode only has 3. We
log a warning if ThermostatMode.AUTO is used for
CurrentHeatingCoolingModeCharacteristic, and return OFF instead.

Fixes #60

ccutrer committed:
fix a pairing issue

make sure we bind the HTTP server to the same address that we told
Bonjour to advertise. otherwise we might end up listening on some rando
interface that may not actually be able to accept connections

ccutrer committed:
bring Characteristic JSON in line with spec

 * events is actually called ev
 * there is no bonjour field

ccutrer committed:
description is optional

don't send it for all the well-known characteristics. Home app doesn't show it,
and it can add up to a significant amount of bytes if you have a lot of
accessories

ccutrer committed:
use short types in JSON

HAP spec allows shortening UUIDs if they're well known Apple ones
doing this drastically reduces response sizes

ccutrer committed:
allow battery status implementations to not actually implement it

i.e. if the implementation is a bridge to another system, so that it can
dynamically provide devices that do or don't have batteries, without wasting
a characteristic and a subscription on it if it doesn't actually have one

ccutrer committed:
batch notifications when possible

often, HomeKit will will change multiple characteristics at once
(on and set brightness of a dimmable bulb, for example) or change
multiple items at once ("hey siri, close all the shades in the house").
instead of sending a separate event message for every characteristic
of every item that changed, batch them all up, and send as a single
message

timcharper commented on src/main/java/com/beowulfe/hap/impl/characteristics/common/ObstructionDetectedCharacteristic.java:
```diff
@@ -18,7 +18,7 @@ public ObstructionDetectedCharacteristic(
       Consumer<HomekitCharacteristicChangeCallback> subscriber,
       Runnable unsubscriber) {
-    super("00000024-0000-1000-8000-0026BB765291", false, true, "Obstruction Detected");
+    super("00000024-0000-1000-8000-0026BB765291", false, true, null);
```
Do we really need to set these to these to null? Or should would modify the API responder to exclude the value?

ccutrer replied:
I can add an override so that we don't have to pass an explicit null, but I think it's useful if you're implementing a custom characteristic to be able to add a description. I don't think Home.app will show the custom characteristic at all, but I know there are 3rd party HomeKit apps that will show all characteristics. You know, a characteristic like https://github.com/timcharper/homebridge-vivint/blob/master/lib/thermostat_characteristics.js#L39 ;)

ccutrer committed:
fix reading encrypted frames that don't line up with network frames

we want to call step again if there's remaining data. pos + toWrite
is what we consumed from this network frame to complete the encrypted
frame. so if it's less than the total network frame (data.length),
then call step again.

andylintner committed:
Merge pull request #64 from ccutrer/communications_drop

fix reading encrypted frames that don't line up with network frames

andylintner committed:
Merge pull request #65 from ccutrer/streamlining2

Some streamlining

andylintner committed:
Merge pull request #67 from ccutrer/listen_correct_interface

fix a pairing issue

ccutrer commented:
Closing for now. most of these commits have been merged into upstream.
